### PR TITLE
Use zod for validation on create endpoint

### LINKED ISSUE #1339: balancer: fix message parsing error -- `invalid type: string "true"`
### Current Behavior

Started seeing this log message from the balancer

```
2024-02-16T14:03:12.261509Z ERROR ott_balancer::balancer: failed to dispatch monolith message e7768db2-da5a-4726-9c07-c98392daf866: invalid type: string "true", expected a boolean at line 1 column 116
```

### Expected Behavior

All messages from monoliths should be valid json, and the balancer should be able to parse them.

### Steps To Reproduce

1. `docker-compose -f docker/with-balancer.docker-compose.yml up -d --scale opentogethertube=10`
2. Watch logs with `docker-compose -f docker/with-balancer.docker-compose.yml logs -f --tail 1`
3. See log message

### Environment

- [ ] This happens on the official site, opentogethertube.com
- [ ] This happens using a self-hosted version.
- [ ] I'm using the docker image.

### Anything else?

I think this has something to do with the `M2BGossip` mesage

### PR DESCRIPTION
**Pull request by Victor-M-Giraldo**
related to #1339

### CODE REVIEW

**Comment by Victor-M-Giraldo** [comment]
I think this works. Zod can also handle validation on the lengths and stuff and there is a library called zod-validation-error which can give nicer errors from what I was reading. 

Let me know if you want any changes/if I missed anything.

File: server/api/room.ts [hunk 1]
```diff
@@ -543,6 +552,7 @@ router.post("/generate", async (req, res, next) => {
 
 router.post("/create", async (req, res, next) => {
 	try {
+		const _validatedBody = createRoomSchema.parse(req.body);
```
Comment (dyc3) [inline server/api/room.ts:555]: This should be done in `createRoom`.

**Review by dyc3 — CHANGES_REQUESTED** [review]
How does this make sure that `createRoomSchema` stays in sync with the type definition `OttApiRequestRoomCreate`? Failing to do so will cause problems in the future.

And make the PR title more descriptive.

**Comment by Victor-M-Giraldo** [comment]
I'll move as much of the validation as I can to Zod. So `OttApiRequestRoomCreate` is like a schema for the request body in the create endpoint? I can see why the two being out of sync would be a problem but I'm not really sure how to account for that. I'll try and find a fix, but I might need more help if I can't.

**Comment by dyc3** [comment]
I remember seeing a `zod.infer` somewhere in the docs.

It might be a good idea to move the zod schemas to ott-common instead, depending on how badly it impacts the bundle size. Vitest also has [a thing to test types](https://vitest.dev/guide/testing-types) that would also work.

**Comment by Victor-M-Giraldo** [comment]
`zod.infer` returns an equivalent type definition from the defined schema. That sounds like what you want I think.

If you find it necessary I can move the zod schemas to ott-common, just give me the confirmation.

I will also take a look at Vitest testing types thing.

**Comment by dyc3** [comment]
Sounds good, but I want to see how badly it affects the bundle size when you build `ott-client` because that affects site performance.

**Comment by Victor-M-Giraldo** [comment]
Is there any difference in functionality if `OttApiRequestRoomCreate` is a type rather than an interface? `z.infer` gives you the type, but is it functionally the same as an interface?

**Comment by dyc3** [comment]
Seems like they are basically the same thing. https://stackoverflow.com/questions/37233735/interfaces-vs-types-in-typescript

**Comment by Victor-M-Giraldo** [comment]
Lint is going to fail here. Adding anything to `common/models/rest-api.ts` makes it fail.

**Comment by Victor-M-Giraldo** [comment]
Little stumped on why the error handler isn't catch the zod error. Also, even when I wrap the schema parsing in a try catch, the request is succeeding in the tests and I can't for the life of me find out why.

If you could push me in the right direction/not straight up give me the answer that would be great.

File: server/api/room.ts [hunk 2]
```diff
@@ -510,6 +492,15 @@ const errorHandler: ErrorRequestHandler = (err: Error, req, res) => {
 					message: err.message,
 				},
 			});
+		} else if (err instanceof ZodError) {
```
Comment (dyc3) [inline server/api/room.ts:495]: Pay close attention to the checks being made and where in the control flow they are being made. This check will never get hit where it is now.

File: server/api/room.ts [hunk 3]
```diff
@@ -119,6 +121,8 @@ const createRoom: RequestHandler<
 	OttResponseBody<OttApiResponseRoomCreate>,
 	OttApiRequestRoomCreate
 > = async (req, res) => {
+	const validatedBody = createRoomSchema.parse(req.body);
```
Comment (dyc3) [inline server/api/room.ts:124]: Also `validatedBody` is currently not used. You should replace usages of `req.body` with the validated body object.

nit: rename `validatedBody` to `body`

Comment (Victor-M-Giraldo) [reply server/api/room.ts:124]: Ah yeah I had it like that before but wasn't sure if it was the same. I'll do that.

File: server/api/room.ts (hunk 2 above)
Comment (Victor-M-Giraldo) [reply server/api/room.ts:495]: Ah yep, got it.

**Comment by Victor-M-Giraldo** [comment]
- Couldn't import reserved room names into `zod-schemas.ts` because of lint errors so I just defined them there and got rid of the old ones.
- I got rid of the error in the async function since we only want to check the name, but now whenever I run the tests I get a bunch of weird output.

**Review by dyc3 — CHANGES_REQUESTED** [review]
Looks good. Just a couple of easy fixes.

**Comment by Victor-M-Giraldo** [comment]
Why do `description` and `queueMode` need to be in this schema? They weren't in the interface for the create endpoint.

**Review by dyc3 — CHANGES_REQUESTED** [review]

**Comment by Victor-M-Giraldo** [comment]
Ah okay, makes sense. 

Forgot to remove those, will do that now.

**Review by dyc3 — APPROVED** [review]

**Pull request merged**